Client/Frontend/Browser/BrowserViewController/OpenSearch/BrowserViewController+OpenSearch.swift

```diff
@@ -0,0 +1,212 @@
+            guard let image = image else {
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                self.present(alert, animated: true) {
+                    self.customSearchEngineButton.state = .enabled
+                }
+                
+                return
+            }
+            
+            NetworkManager().downloadResource(with: url).uponQueue(.main) { [weak self] response in
+                guard let openSearchEngine =
+                        OpenSearchParser(pluginMode: true).parse(response.data, referenceURL: referenceURL, image: image) else {
+                    return
+                }
+                
+                self?.addSearchEngine(openSearchEngine)
+            }
+        })
+    }
+    
+    func addSearchEngine(_ engine: OpenSearchEngine) {
+        let alert = ThirdPartySearchAlerts.addThirdPartySearchEngine(engine) { alert in
+            do {
+                try self.profile.searchEngines.addSearchEngine(engine)
+                
+                let toast = SimpleToast()
+                toast.showAlertWithText(Strings.thirdPartySearchEngineAdded, bottomContainer: self.webViewContainer)
+                
+                self.customSearchEngineButton.state = .disabled
+            } catch {
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                self.present(alert, animated: true) {
+                    self.customSearchEngineButton.state = .enabled
+                }
+            }
+        }
+
+        self.present(alert, animated: true, completion: {})
+    }
+}
+
+// MARK: - KeyboardHelperDelegate
+
+extension BrowserViewController: KeyboardHelperDelegate {
+    func keyboardHelper(_ keyboardHelper: KeyboardHelper, keyboardWillShowWithState state: KeyboardState) {
+        keyboardState = state
+        updateViewConstraints()
+
+        UIView.animate(withDuration: state.animationDuration) {
+            UIView.setAnimationCurve(state.animationCurve)
+            self.alertStackView.layoutIfNeeded()
+        }
+
+        guard let webView = tabManager.selectedTab?.webView else { return }
+
+        self.evaluateWebsiteSupportOpenSearchEngine(webView)
+    }
+
+    func keyboardHelper(_ keyboardHelper: KeyboardHelper, keyboardWillHideWithState state: KeyboardState) {
+        keyboardState = nil
+        updateViewConstraints()
+        //If the searchEngineButton exists remove it form the keyboard
+        if let buttonGroup = customSearchBarButton?.buttonGroup {
+            buttonGroup.barButtonItems = buttonGroup.barButtonItems.filter { $0 != customSearchBarButton }
+            customSearchBarButton = nil
+        }
+
+        if self.customSearchEngineButton.superview != nil {
+            self.customSearchEngineButton.removeFromSuperview()
+        }
+
+        UIView.animate(withDuration: state.animationDuration) {
+            UIView.setAnimationCurve(state.animationCurve)
```

Comment:
Instead fo `setAnimationCurve` will `animate(withDuration:delay:options:animations:completion:)` work if you supply the `curve` as `options` parameter?

Reply:
It can work yes but in that case you will need to use a convenience method to get the animation option according to which type of curve is used. 

```
func animationOption(for curve: UIViewAnimationCurve) -> UIViewAnimationOptions {
	switch (curve) {
	case .easeInOut:
		return .curveEaseInOut
	case .easeIn:
		return .curveEaseIn
	case .easeOut:
		return .curveEaseOut
	case .linear:
		return .curveLinear
	}
}
```

So I believe their way for handling this is easier